### PR TITLE
Implement generate_tests macro for generated test cases

### PR DESCRIPTION
**Pull request by kontheocharis**
This can be found in the `hash_utils_testing_macros` crate.

Test cases are generated based on a given test folder path. Each generated test corresponds to the full path of each leaf node of the test folder structure. Each test case runs a predefined function which is called with some contextual data (i.e. path of test case).

See docs of `generate_tests` for more information.

### CODE REVIEW

**Comment by kontheocharis** [comment]
Failing due to nightly features, opened PR #53 to change this.

File: compiler/hash-utils/testing-macros/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "hash-utils-testing-macros"
+version = "0.1.0"
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (feds01) [inline compiler/hash-utils/testing-macros/Cargo.toml:6]: unnecessary comment?

File: compiler/hash-utils/testing-macros/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+/// of the test folder structure. The generator will stop generating sub-tests for a subfolder once
+/// any file in that subfolder matches `TEST_PATTERN`. For example, if the test folder structure is
+/// such that some test folder `test_foo/` contains files `case.hash` and `output`, specifying
+/// either of `"output"` or `"case.hash"` as `TEST_PATTERN` will result in a test case being
+/// generated for `test_foo`, but no further (i.e. not `test_foo_output` or `test_foo_case_hash`).
+///
+/// - `TEST_FN` must be an expression of type [`TestingFn`](hash_utils::testing::TestingFn). Every
+/// generated test case body will invoke this function with the appropriate
+/// [`TestingInput`](hash_utils::testing::TestingInput). Within this function, the actual test
+/// logic should be written.
```
Comment (feds01) [inline compiler/hash-utils/testing-macros/src/lib.rs:140]: Incredible docs

**Review by feds01 — APPROVED** [review]
LGTM

File: compiler/hash-utils/testing-macros/src/lib.rs (hunk 2 above)
Comment (kontheocharis) [reply compiler/hash-utils/testing-macros/src/lib.rs:140]: thank you kind sir

**Pull request merged**